boxtree/tree_build.py

```diff
@@ -46,6 +46,207 @@ class MaxLevelsExceeded(RuntimeError):
     pass
 
 
+# {{{ utils for tree of boxes
+
+def resized_array(arr, new_size):
+    """Return a resized copy of the array. The new_size is a scalar which is
+    applied to the last dimension.
+    """
+    old_size = arr.shape[-1]
+    prefix = (slice(None), ) * (arr.ndim - 1)
+    if old_size >= new_size:
+        return arr[prefix + (slice(new_size), )].copy()
+    else:
+        new_shape = list(arr.shape)
+        new_shape[-1] = new_size
+        new_arr = np.zeros(new_shape, arr.dtype)
+        new_arr[prefix + (slice(old_size), )] = arr
+        return new_arr
+
+
+def vec_of_signs(dim, i):
+    """The sign vector is obtained by converting i to a dim-bit binary.
+    """
+    # e.g. bin(10) = '0b1010'
+    binary_digits = [int(bd) for bd in bin(i)[2:]]
+    n = len(binary_digits)
+    assert n <= dim
+    return np.array([0]*(dim-n) + binary_digits) * 2 - 1
+
+
+# {{{ refine/coarsen a tree of boxes
+
+def refined(tob, refine_flags):
+    """
+    Make a refined copy of tob.
+    """
+    return refine_and_coarsened(tob, refine_flags, None)
+
+
+def uniformly_refined(tob):
+    refine_flags = np.zeros(tob.nboxes, bool)
+    refine_flags[tob.get_leaf_flags()] = 1
+    return refined(tob, refine_flags)
+
+
+def coarsened(tob, coarsen_flags):
+    return refine_and_coarsened(tob, None, coarsen_flags)
+
+
+def refine_and_coarsened(tob, refine_flags=None, coarsen_flags=None):
```

Comment:
A lot of the functionality here doesn't seem to be included in the sphinx docs.

Reply:
d636ec9b1b380890631e21a4ab3df33867c98e67